lib/Platform/Intl/PlatformIntlICU.cpp

```diff
@@ -120,40 +419,782 @@ double Collator::compare(
+      skeleton += u"y";
+    else if (year_ == eTwoDigit)
+      skeleton += u"yy";
+  }
+
+  if (month_.has_value()) {
+    if (month_ == eTwoDigit)
+      skeleton += u"MM";
+    else if (month_ == eNumeric)
+      skeleton += u'M';
+    else if (month_ == eNarrow)
+      skeleton += u"MMMMM";
+    else if (month_ == eShort)
+      skeleton += u"MMM";
+    else if (month_ == eLong)
+      skeleton += u"MMMM";
+  }
+
+  if (day_.has_value()) {
+    if (day_ == eNumeric)
+      skeleton += u"d";
+    else if (day_ == eTwoDigit)
+      skeleton += u"dd";
+  }
+
+  if (hour_.has_value()) {
+    if (hourCycle_ == u"h12") {
+      if (hour_ == eNumeric)
+        skeleton += u"h";
+      else if (hour_ == eTwoDigit)
+        skeleton += u"hh";
+    } else if (hourCycle_ == u"h24") {
+      if (hour_ == eNumeric)
+        skeleton += u"k";
+      else if (hour_ == eTwoDigit)
+        skeleton += u"kk";
+    } else if (hourCycle_ == u"h23") {
+      if (hour_ == eNumeric)
+        skeleton += u"k";
+      else if (hour_ == eTwoDigit)
+        skeleton += u"KK";
+    } else {
+      if (hour_ == eNumeric)
+        skeleton += u"h";
+      else if (hour_ == eTwoDigit)
+        skeleton += u"HH";
+    }
+  }
+
+  if (minute_.has_value()) {
+    if (minute_ == eNumeric)
+      skeleton += u"m";
+    else if (minute_ == eTwoDigit)
+      skeleton += u"mm";
+  }
+
+  if (second_.has_value()) {
+    if (second_ == eNumeric)
+      skeleton += u"s";
+    else if (second_ == eTwoDigit)
+      skeleton += u"ss";
+  }
+
+  UErrorCode status = U_ZERO_ERROR;
+  std::u16string bestpattern;
+  int32_t patternLength;
+
+  UDateTimePatternGenerator *dtpGenerator = udatpg_open(&locale8_[0], &status);
+  patternLength = udatpg_getBestPatternWithOptions(
+      dtpGenerator,
+      &skeleton[0],
+      -1,
+      UDATPG_MATCH_ALL_FIELDS_LENGTH,
+      nullptr,
+      0,
+      &status);
+
+  if (status == U_BUFFER_OVERFLOW_ERROR) {
+    status = U_ZERO_ERROR;
+    bestpattern.resize(patternLength);
+    udatpg_getBestPatternWithOptions(
+        dtpGenerator,
+        &skeleton[0],
+        skeleton.length(),
+        UDATPG_MATCH_ALL_FIELDS_LENGTH,
+        &bestpattern[0],
+        patternLength,
+        &status);
+  }
+
+  // if timezone is specified, use that instead, else use default
+  if (!timeZone_.empty()) {
+    const UChar *timeZoneRes =
+        reinterpret_cast<const UChar *>(timeZone_.c_str());
+    int32_t timeZoneLength = timeZone_.length();
+    return udat_open(
+        UDAT_PATTERN,
+        UDAT_PATTERN,
+        &locale8_[0],
+        timeZoneRes,
+        timeZoneLength,
+        &bestpattern[0],
+        patternLength,
+        &status);
+  } else {
+    return udat_open(
+        UDAT_PATTERN,
+        UDAT_PATTERN,
+        &locale8_[0],
+        nullptr,
+        -1,
+        &bestpattern[0],
+        patternLength,
+        &status);
+  }
+}
+
+vm::CallResult<std::u16string> DateTimeFormatICU::getDefaultHourCycle(
+    vm::Runtime &runtime) {
```

Comment:
The default hour cycle depends on the user specified locale, so we shouldn't be using the default locale to compute this.

Reply:
we use `locale8_`, which comes from `resolveLocale`, to compute it